module/zfs/zap_micro.c

```diff
@@ -149,43 +149,96 @@ zap_normalize(zap_t *zap, const char *name, char *namenorm, int normflags)
 boolean_t
 zap_match(zap_name_t *zn, const char *matchname)
 {
+	boolean_t res = B_FALSE;
 	ASSERT(!(zap_getflags(zn->zn_zap) & ZAP_FLAG_UINT64_KEY));
 
 	if (zn->zn_matchtype & MT_NORMALIZE) {
-		char norm[ZAP_MAXNAMELEN];
+		size_t namelen = zn->zn_normbuf_len;
+		char *norm;
 
+		norm = kmem_alloc(namelen, KM_SLEEP);
```

Comment:
It seems `zap_match()` is called in a loop over potentially many other entries. Allocation for might be expensive.  Sure using stack previously was a cheating, but this might need some more thinking.

Reply:
I can go back to the original one.